shithead_game_server/src/lobby/mod.rs

```diff
@@ -289,50 +332,49 @@ impl Lobby {
         client_id: ClientId,
         clicked_card_location: ClickedCardLocation,
     ) -> Result<(), GameServerError> {
-        let current_turn_player_id = self
-            .current_turn
-            .as_ref()
-            .ok_or(GameServerError::GameHasntStartedYet)?
-            .player_id();
-
-        if current_turn_player_id != client_id {
-            return Err(GameServerError::NotYourTurn);
-        }
-
-        let player = self
-            .player_list
-            .get_player(client_id)
-            .ok_or(GameServerError::NotInALobby)?;
-
-        match clicked_card_location {
-            ClickedCardLocation::Trash => {
-                // if the player can place any of his cards in the trash, then he is not allowed to
-                // take the trash
-                if player
-                    .what_cards_can_be_placed_on(self.trash_top_card_rank())
-                    .next()
-                    .is_some()
-                {
-                    return Err(GameServerError::CantTakeTrashBecauseSomeCardsCanBePlayed);
+        match &mut self.state {
+            LobbyState::Waiting => return Err(GameServerError::GameHasntStartedYet),
+            LobbyState::ChoosingTop3(_) => {
+                let player = self
+                    .data
+                    .player_list
+                    .get_player_mut(client_id)
+                    .ok_or(GameServerError::NotInALobby)?;
+                match clicked_card_location {
+                    ClickedCardLocation::FromCardsInHand { card_index } => {
+                        // the player clicked a card from his hand during the top 3 selection, move
+                        // that card to his top 3 cards.
+                        let card_index = card_index as usize;
+                        if card_index >= player.cards_in_hand.len() {
+                            return Err(GameServerError::NoSuchCard);
+                        }
```

Comment:
This check will get repetitive. Maybe you should create a method for that?

Reply:
Creating a method won't make the code much shorter or cleaner since it is already pretty concise, so i think i won't do it unless i'll see that it gets really repetitive.